### PR TITLE
[TTAHUB-3097] Implement transactions for worker processes

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change
This update introduces transaction handling for worker processes in the Head Start TTA Data Platform. The main changes include:

1. Implementing transaction management in worker scripts to ensure data consistency.
2. Enhancing error logging for detailed information about errors and database connection states.
3. Updating tests to align with the new transaction management and error handling.

## How to test

1. Run Unit Tests: Execute all unit tests to ensure they pass with the new transaction handling logic.
2. Simulate Worker Processes: Trigger worker processes and monitor logs for detailed error messages.
3. Check Database Consistency: Verify that data remains consistent and correctly rolled back in case of errors during worker processes. 

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-3097

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/lib/uuidConverter.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { v5 as uuidv5, validate as validateUUID, v4 as uuidv4 } from 'uuid';
+
+const UUID_NAMESPACE = uuidv4(); // You can use a fixed namespace UUID
+
+/**
+ * Converts a given job ID to a UUID format.
+ * @param jobId - The job ID to be converted.
+ * @returns The UUID formatted job ID.
+ */
+export default function convertToUUID(jobId: string): string {
```
Comment (thewatermethod) [inline src/lib/uuidConverter.ts:10]: This isn't used anywhere, despite being imported

**Comment by GarrettEHill** [comment]
@thewatermethod @nvms @AdamAdHocTeam @kcirtapfromspace @kryswisnaskas @hardwarehuman 
Its all passing now, ready for review... again

**Review by nvms — COMMENTED** [review]
Let's simplify `transactionWrapper` a little bit:

```typescript
export default function transactionWrapper(originalFunction, context = '') {
  return async function wrapper(req, res, next) {
    const startTime = Date.now();
    try {
      // eslint-disable-next-line @typescript-eslint/return-await
      return await sequelize.transaction(async () => {
        try {
          await addAuditTransactionSettings(sequelize, null, null, 'transaction', originalFunction.name);
          const result = await originalFunction(req, res, next);
          const duration = Date.now() - startTime;
          auditLogger.info(`${originalFunction.name} ${context} execution time: ${duration}ms`);
          removeFromAuditedTransactions();
          return result;
        } catch (err) {
          auditLogger.error(`Error executing ${originalFunction.name} ${context}: ${err.message}`);
          throw err;
        }
      });
    } catch (err) {
      return handleErrors(req, res, err, logContext);
    }
  };
}
```

So we don't have to do that `error = err` thing, which isn't needed. Also the result can be returned inside the try, so it doesn't need to be in the upper scope. Just a bit easier to read IMO.

File: src/services/requestErrors.js [hunk 2]
```diff
@@ -8,24 +6,28 @@ export default async function createRequestError({
   responseBody = 'N/A',
   responseCode = 'N/A',
 }) {
+  // eslint-disable-next-line global-require
+  const { RequestErrors } = require('../models');
```
Comment (thewatermethod) [inline src/services/requestErrors.js:10]: why is this necessary now?

File: src/workers/transactionWrapper.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  originalFunction: (job: Job) => Promise<any>,
+  context = '',
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+) => async (job: Job): Promise<any> => {
+  const startTime = Date.now();
+  return httpContext.ns.runPromise(async () => {
+    httpContext.set('loggedUser', job.referenceData.userId);
+    httpContext.set('impersonationUserId', job.referenceData.impersonationUserId);
+    httpContext.set('sessionSig', job.id); // TODO: what value should be used here
```
Comment (thewatermethod) [inline src/workers/transactionWrapper.ts:25]: is this TODO still applicable?

File: src/lib/apiErrorHandler.js [hunk 4]
```diff
@@ -141,3 +118,110 @@ export default async function handleErrors(req, res, error, logContext) {
...
+    operation = 'SequelizeError';
+    label = 'Sequelize error';
+  } else {
+    operation = 'UNEXPECTED_ERROR';
+    label = 'UNEXPECTED ERROR';
+  }
+
+  if (error instanceof Sequelize.ConnectionError
+    || error instanceof Sequelize.ConnectionAcquireTimeoutError) {
+    logger.error(`${logContext.namespace} Connection Pool: ${JSON.stringify(sequelize.connectionManager.pool)}`);
```
Comment (thewatermethod) [inline src/lib/apiErrorHandler.js:189]: same as above

File: src/services/requestErrors.js (hunk 2 above)
Comment (GarrettEHill) [reply src/services/requestErrors.js:10]: I don't know specifically which change made this required, but there seems to be a recursive include that was making it non-fictional without this change.

File: src/lib/apiErrorHandler.js [hunk 5]
```diff
@@ -86,24 +75,14 @@ export const handleError = async (req, res, error, logContext) => {
     label = 'UNEXPECTED ERROR';
   }
 
-  // eslint-disable-next-line max-len
-  if (error instanceof Sequelize.ConnectionError || error instanceof Sequelize.ConnectionAcquireTimeoutError) {
-    logger.error(`${logContext.namespace} Connection Pool: ${JSON.stringify(sequelize.connectionManager.pool)}`);
+  if (error instanceof Sequelize.ConnectionError
+    || error instanceof Sequelize.ConnectionAcquireTimeoutError) {
+    logger.error(`${logContext.namespace} Connection Pool: ${JSON.stringify(sequelize?.connectionManager?.pool)}`);
```
Comment (kryswisnaskas) [inline src/lib/apiErrorHandler.js:80]:
```suggestion
      const pool = sequelize.connectionManager.pool;
  const usedConnections = pool.used.length;
  const waitingConnections = pool.pending.length;

  logger.error(`${logContext.namespace} Connection Pool: Used Connections - ${usedConnections}, Waiting Connections - ${waitingConnections}`);
```

[the remark this answers is not in the document]
Comment (kryswisnaskas) [reply src/lib/apiErrorHandler.js:80]: How about this instead?

**Comment by thewatermethod** [comment]
I'd grab Jon's changes to the sequelizeConnectionError stuff, or wait until that's merged into main, so we don't merge anything that throws errors in the error handler

File: src/lib/apiErrorHandler.js [hunk 6]
```diff
@@ -13,37 +13,28 @@ import { sequelize } from '../models';
...
   }
+  if (!error) {
+    return 0;
+  }
 
   try {
-    // Prepare the response body for storage
-    const responseBody = typeof error === 'object'
-      && error !== null ? { ...error, errorStack: error?.stack } : error;
+    const responseBody = typeof error === 'object' && error !== null
```
Comment (AdamAdHocTeam) [inline src/lib/apiErrorHandler.js:28]: if we check if(!error) above do we need to check error !== null here?

File: src/lib/apiErrorHandler.js [hunk 7]
```diff
@@ -13,37 +13,28 @@ import { sequelize } from '../models';
...
-      ...(req.query
-        && typeof req.query === 'object'
-        && Object.keys(req.query).length > 0
-        && { query: req.query }),
+      ...(req.body && typeof req.body === 'object' && Object.keys(req.body).length > 0 && { body: req.body }),
+      ...(req.params && typeof req.params === 'object' && Object.keys(req.params).length > 0 && { params: req.params }),
+      ...(req.query && typeof req.query === 'object' && Object.keys(req.query).length > 0 && { query: req.query }),
     };
 
-    // Create a request error in the database and get its ID
```
Comment (AdamAdHocTeam) [inline src/lib/apiErrorHandler.js:46]: Not a fan of comments?

File: src/lib/apiErrorHandler.js [hunk 8]
```diff
@@ -144,3 +121,113 @@ export default async function handleErrors(req, res, error, logContext) {
...
+    operation !== 'SequelizeError'
+    && process.env.SUPPRESS_ERROR_LOGGING
+    && process.env.SUPPRESS_ERROR_LOGGING.toLowerCase() === 'true'
+  ) {
+    return 0;
+  }
+
+  try {
+    const responseBody = typeof error === 'object' && error !== null
+      ? { ...error, errorStack: error?.stack }
```
Comment (AdamAdHocTeam) [inline src/lib/apiErrorHandler.js:144]: Wouldn't ...error also contain error.stack. Why do we need to put it in the property errorStack?

File: src/lib/apiErrorHandler.js [hunk 9]
```diff
@@ -144,3 +121,113 @@ export default async function handleErrors(req, res, error, logContext) {
...
+ */
+export const handleWorkerError = async (job, error, logContext) => {
+  if (process.env.NODE_ENV === 'development') {
+    logger.error(error);
+  }
+
+  let operation;
+  let label;
+
+  if (error instanceof Sequelize.Error) {
```
Comment (AdamAdHocTeam) [inline src/lib/apiErrorHandler.js:182]: why do we check instanceof Sequelize.Error here but above we do operation !== 'SequelizeError'.

File: src/lib/apiErrorHandler.js [hunk 10]
```diff
@@ -144,3 +121,113 @@ export default async function handleErrors(req, res, error, logContext) {
...
+};
+
+/**
+ * Handles any unexpected errors in a worker error handler catch block.
+ * @param {Object} job - The job object.
+ * @param {Error} error - The error object.
+ * @param {Object} logContext - The context for logging.
+ */
+export const handleUnexpectedWorkerError = (job, error, logContext) => {
+  logger.error(`${logContext.namespace} - Unexpected error in catch block - ${error}`);
```
Comment (AdamAdHocTeam) [inline src/lib/apiErrorHandler.js:218]: Will call stack be visible here?

File: src/services/scanQueue.js [hunk 11]
```diff
@@ -35,7 +40,8 @@ const processScanQueue = () => {
   scanQueue.on('failed', onFailedScanQueue);
   scanQueue.on('completed', onCompletedScanQueue);
   increaseListeners(scanQueue);
-  scanQueue.process((job) => processFile(job.data.key));
+  const process = (job) => processFile(job.data.key);
+  scanQueue.process(transactionQueueWrapper(process));
```
Comment (AdamAdHocTeam) [inline src/services/scanQueue.js:44]: Do we need to pass a second param here to the transactionQueueWrapper like in S3 and Resource queue's?

File: src/lib/apiErrorHandler.js (hunk 8 above)
Comment (GarrettEHill) [reply src/lib/apiErrorHandler.js:144]: just matching what the original code had

File: src/lib/apiErrorHandler.js (hunk 9 above)
Comment (GarrettEHill) [reply src/lib/apiErrorHandler.js:182]: just matching original logic

File: src/lib/apiErrorHandler.js (hunk 10 above)
Comment (GarrettEHill) [reply src/lib/apiErrorHandler.js:218]: i think so

File: src/services/scanQueue.js (hunk 11 above)
Comment (GarrettEHill) [reply src/services/scanQueue.js:44]: the scan queue never passed a key, so there is no associated value to use as a context

**Review by thewatermethod — APPROVED** [review]

**Review by nvms — APPROVED** [review]

**Pull request merged**